src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,101 @@
-    @Override
-    public void addAll(List<T> list) {
-
-    }
-
-    @Override
-    public T get(int index) {
-        return null;
-    }
-
-    @Override
-    public void set(T value, int index) {
-
-    }
-
-    @Override
-    public T remove(int index) {
-        return null;
-    }
-
-    @Override
-    public T remove(T element) {
-        return null;
-    }
-
-    @Override
-    public int size() {
-        return 0;
-    }
-
-    @Override
-    public boolean isEmpty() {
-        return false;
-    }
-}
+package core.basesyntax;
+
+import java.util.NoSuchElementException;
+
+public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_FACTOR = 1.5;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    @Override
+    public void add(T value) {
+        grow();
+        elementData[size++] = value;
+    }
+
+    @Override
+    public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index " + index
+                    + " out of list size " + size);
+        }
+        grow();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
+    }
+
+    @Override
+    public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+    }
+
+    @Override
+    public T get(int index) {
+        checkIndex(index);
+        return elementData[index];
+    }
+
+    @Override
+    public void set(T value, int index) {
+        checkIndex(index);
+        elementData[index] = value;
+    }
+
+    @Override
+    public T remove(int index) {
+        checkIndex(index);
+        T removeElement = elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index,
+                elementData.length - index - 1);
+        size--;
+        return removeElement;
+    }
+
+    @Override
+    public T remove(T element) {
+        for (int i = 0; i < elementData.length; i++) {
+            if (elementData[i] == element || elementData[i] != null
+                    && elementData[i].equals(element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("element " + element + " not found");
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index " + index
+                    + " out of list size " + size);
+        }
+    }
+
+    private void grow() {

```

Comment:
name implies that grow is expected to happen every time, it would be more precise to name it something like growIfNeed to reflect that the check is happening inside, and grow is conditional on it